R/run-nmtran.R

```diff
@@ -0,0 +1,330 @@
+    }else if(presort_status_val != 0){
+      return(
+        list(
+          nmtran_model = file.path(run_dir, mod_path),
+          run_dir = run_dir,
+          status = "nmtran_presort failed. See errors.",
+          status_val = presort_status_val,
+          output_lines = presort.p$stdout
+        )
+      )
+    }
+
+    # Write the output to tempzzzz1 control stream
+    writeLines(presort.p$stdout, con = file.path(run_dir, "tempzzzz1.ctl"))
+  }
+
+  # Run NM-TRAN
+  stdin_file <- ifelse(run_presort, "tempzzzz1.ctl", mod_path)
+  nmtran.p <- processx::run(
+    command = nmtran_exe, wd = run_dir, args = cmd_args,
+    stdout = "|", stdin = file.path(run_dir, stdin_file),
+    stderr_to_stdout = TRUE, error_on_status = FALSE
+  )
+
+  # Assign status
+  status_val <- nmtran.p$status
+  if(is.na(status_val)){
+    rlang::abort("NM-TRAN terminated unexpectedly")
+  }else if(status_val == 0){
+    status <- "NM-TRAN successful"
+  }else{
+    status <- "NM-TRAN failed. See errors."
+  }
+
+  # Tabulate NM-TRAN results
+  nmtran_results <- list(
+    nmtran_model = stdin_file,
+    run_dir = run_dir,
+    status = status,
+    status_val = status_val,
+    output_lines = nmtran.p$stdout
+  )
+
+  return(nmtran_results)
+}
+
+
+#' Helper for finding and reading in a `bbi.yaml` file given a `bbi_nonmem_model`
+#' object or explicit `.config_path`.
+#' @inheritParams run_nmtran
+#' @noRd
+get_bbi_yaml_path <- function(.mod = NULL, .config_path = NULL){
+  if(!is.null(.mod)){
+    check_model_object(.mod, "bbi_nonmem_model")
+    model_dir <- get_model_working_directory(.mod)
+  }
+  bbi_yaml_path <- .config_path %||% file.path(model_dir, "bbi.yaml")
+
+  if(!file_exists(bbi_yaml_path)){
+    if(is.null(.config_path)){
+      msg <- c(
+        "No bbi configuration was found in the execution directory.",
+        "i" = "Please run `bbi_init()` with the appropriate directory to continue."
+      )
+    }else{
+      msg <- glue("No bbi configuration was found at {.config_path}")
+    }
+    rlang::abort(msg)
+  }
+
+  return(bbi_yaml_path)
+}
+
+
+#' Parse `.bbi_args` and return the three expected `nmfe_options` for `NM-TRAN`
+#'  in the correct format
+#' @inheritParams run_nmtran
+#' @param nmfe_options named list of nmfe options defined in a `bbi.yaml` file.
+#' @noRd
+parse_nmtran_args <- function(
+    .mod,
+    .bbi_args = NULL,
+    nmfe_options = NULL
+){
+
+  nmfe_args_def <- list(prdefault = FALSE, tprdefault = FALSE, maxlim = 2)
+
+  # Combine with any options stored in model yaml, preferring .bbi_args
+  .nmfe_args <- parse_args_list(.bbi_args, .mod[[YAML_BBI_ARGS]])
+
+  # Combine with nmfe options stored in bbi.yaml, preferring .nmfe_args
+  .nmfe_args <- parse_args_list(.nmfe_args, nmfe_options)
+
+  # Check provided args
+  check_bbi_args(.nmfe_args)
+
+  # Combine with and filter to default nmfe_options
+  .nmfe_args <- parse_args_list(.nmfe_args, nmfe_args_def)
+  .nmfe_args <- .nmfe_args[names(nmfe_args_def)]
```

Comment:
This logic is looking pretty good. Aside from yaml parsing differences between Go and R (they exist but I very much think we should let that be), I spot one place where the behavior would diverge. A `FALSE` value for `.bbi_arg` has no effect. It does not override the value from `bbi.yaml`.

For example, a run with a `bbi.yaml` that has `prdefault: true` and with `list(prdefault = FALSE)` for bbi args, the underlying bbi will end up passing `-prdefault`:

```
$ grep prdef inst/model/nonmem/basic/1/1.sh
/opt/NONMEM/nm75/run/nmfe75 1.ctl  1.lst  -parafile=1.pnm -prdefault -maxlim=2
```

It is admittedly confusing behavior, but having `run_nmtran` diverge would just add to the confusion.

Reply:
Just to make sure we had the same setup, when you passed `prdefault: true` in the `bbi.yaml`, did you do so under `nmfe_options`, or as a `bbi_arg`? I only looked at `nmfe_options` since that's where `maxlim` is specified. This is a bit confusing given `print_bbi_args` documentation, which to me suggests these 3 nmfe options can be treated as `bbi_args` too.

Reply:
> when you passed `prdefault: true` in the `bbi.yaml`, did you do so under `nmfe_options`, or as a `bbi_arg`?

Yep, the nmfe options are specified under `nmfe_options` in `bbi.yaml`.

```
nmfe_options:
  prdefault: true
```

> This is a bit confusing given `print_bbi_args` documentation, which to me suggests these 3 nmfe options can be treated as `bbi_args` too.

They are bbi arguments, and you can specify them via the `.bbi_args` argument to `submit_model` or the `bbi_args` field in the model yaml.  At the configuration level, bbi groups the nmfe ones under their own "section".

When you call bbi like so

```
$ bbi nonmem run local --prdefault 1.mod
```

it relays that to the nmfe call:

```
$ grep prdefault 1/1.sh
/opt/NONMEM/nm75/run/nmfe75 1.mod  1.lst  -prdefault -maxlim=2
```

Reply:
@kyleam right. My anxiety stems from the idea that users could do this though:
```r
bbi_init(.bbi_args = list(prdefault = TRUE))
```
which would _not_ place those arguments under `nmfe_options`. In that case `run_nmtran` and `bbi` would both ignore those arguments if im not mistaken. If that's the case, the only way to really pass these options would be via `.bbi_args` in `run_nmtran` or `submit_model`, or as part of the model object:
```r
mod <-  add_bbi_args(mod, list(prdefault = TRUE))
```

I dont see how these nmfe options could be specified via the `bbi.yaml` without manual intervention (adding directly under `nmfe_options`)

Reply:
> [bbi_init] which would _not_ place those arguments under `nmfe_options`.

Good catch.  (I'm not sure I was even aware `bbi_init` supported writing user-specified bbi arguments to `bbi.yaml`.)  Yes, it looks like `bbi_init` puts nmfe arguments in the wrong spot.  

> In that case `run_nmtran` and `bbi` would both ignore those arguments if im not mistaken.

Yes.  `run_nmtran` is right to follow `bbi`, since that is what is in effect for the actual run.

> I dont see how these nmfe options could be specified via the `bbi.yaml` without manual intervention (adding directly under `nmfe_options`)

The user is free to modify it, so we need to account for it in this handling.

As for the `bbi_init` bug, it's orthogonal to this PR.  Would you mind filing an issue for it?

Reply:
[Relevant issue](https://github.com/metrumresearchgroup/bbr/issues/714)